File: c++/0007-reverse-integer.cc
#include <gtest/gtest.h>

#include <limits>

namespace reverseInteger {

int solve(int x) {
  int result = 0;

  while (x != 0) {
    // Check for overflow before multiplying
    if (result > std::numeric_limits<int>::max() / 10 ||
        (result == std::numeric_limits<int>::max() / 10 && x % 10 > 7)) {
      return 0;
    }
    if (result < std::numeric_limits<int>::min() / 10 ||
        (result == std::numeric_limits<int>::min() / 10 && x % 10 < -8)) {
      return 0;
    }

    result = result * 10 + x % 10;
    x /= 10;
  }

  return result;
}

}  // namespace reverseInteger

TEST(ReverseInteger, TestCases) {
  EXPECT_EQ(reverseInteger::solve(123), 321);
  EXPECT_EQ(reverseInteger::solve(-123), -321);
  EXPECT_EQ(reverseInteger::solve(120), 21);
}

File: swift/Sources/LeetCode/0007ReverseInteger.swift
public func reverseInteger(_ x: Int) -> Int {
    var x = x
    var result = 0
    
    while x != 0 {
        if result > Int.max / 10 || (result == Int.max / 10 && x > 7) {
            return 0
        }
        if result < Int.min / 10 || (result == Int.min / 10 && x < -8) {
            return 0
        }
        result = result * 10 + x % 10
        x /= 10
    }
    
    return result
}
